3.0 config engine

ADmad opened the pull request:
I extracted the common code required for file based config engines into a trait for DRYness.

Also the `$key` passed to `read()` / `dump()`should no longer contain file extension. Keys should be engine agnostic and the support for keys with extension was only kept for backwards compatibility with 1.x when config engines were introduced in 2.x.

ADmad committed:
Create FileConfigTrait to avoid duplicate code in file based config engines.

ADmad committed:
Make code DRYer.

ADmad committed:
Fix tests.

ADmad committed:
Add JsonConfig configure engine.

ADmad commented:
I added JsonConfig engine suggested in #5563.

markstory commented on src/Core/Configure/Engine/JsonConfig.php:
```diff
@@ -0,0 +1,82 @@
+    public function read($key)
+    {
+        $file = $this->_getFilePath($key, true);
+        return (array)json_decode(file_get_contents($file), true);
```
What if there is an error, you might want to raise an exception. For example if there is a syntax error in the JSON file.

ADmad replied:
Will do. How about empty files? Should an error be thrown or just return empty array? The latter would be probably be better.

markstory commented on tests/TestCase/Core/Configure/Engine/IniConfigTest.php:
```diff
@@ -270,8 +270,8 @@ public function testDump()
     {
         $engine = new IniConfig(TMP);
-        $engine->dump('test.ini', $this->testData);
-        $result = $engine->read('test.ini');
```
Why did the extension get removed in these tests?

ADmad replied:
Please check the second paragraph of issue description.

markstory replied:
Ah ok.

ADmad committed:
Throw an exception if json decoding fails.

ADmad commented:
@markstory I added the exception in case of json decoding failing.

lorenzo commented on src/Core/Configure/Engine/JsonConfig.php:
```diff
@@ -0,0 +1,89 @@
+        $file = $this->_getFilePath($key, true);
+
+        $values = json_decode(file_get_contents($file), true);
+        if (json_last_error() !== JSON_ERROR_NONE) {
```
I would also throw exception if it is not an array

markstory replied:
Would be good to give the json error too.

ADmad replied:
@lorenzo Done

@markstory `json_last_error_msg()` is only available since PHP 5.5. Should I add a function to basics.php like shown here http://php.net/manual/en/function.json-last-error-msg.php#113243 ?  It would be useful in showing specific error in JsonView too.

markstory replied:
Converting the error code into a message might be a useful thing to do in both places like you said.

ADmad committed:
Throw exception if json file is empty or result of decoding is not array.

ADmad committed:
Show specific error message in exception if json_decode() fails.

Added json_last_error_msg() as fallback as it's not available in PHP 5.4 and below.

ADmad committed:
Fix CS error

ADmad commented:
@markstory I added the method and display the error message in exception. Will take care of showing the error message in JsonView separately.

markstory commented on src/Core/Configure/Engine/JsonConfig.php:
```diff
@@ -0,0 +1,100 @@
+            throw new Exception(sprintf(
+                'Decoding JSON config file "%s.json" did not return any array',
+                $key,
+                json_last_error_msg()
```
There is no placeholder for this value.

dereuromark replied:
`%s` should be `{0}` in 3.x

ADmad replied:
@dereuromark This isn't a translated string.

markstory commented on src/Core/Configure/Engine/JsonConfig.php:
```diff
@@ -0,0 +1,100 @@
+        }
+        if (!is_array($values)) {
+            throw new Exception(sprintf(
+                'Decoding JSON config file "%s.json" did not return any array',
```
'did not return an array'

ADmad committed:
Remove extraneous argument to sprintf call.

ADmad commented:
@markstory Fixed the sprintf call

ADmad commented:
Anything else needs to be addressed here?

lorenzo commented:
:clap:

berarma commented:
This change produces this error for me:
> Error: The application is trying to load a file from the app_form plugin.
> Make sure your plugin app_form is in the /home/bernat/work/new-roomsday3/plugins/ directory and was loaded.

The file config/app_form.php exists but it's only used in AppController to load new templates in FormHelper as suggested in the book.

ADmad commented:
@berarma  Make sure you have specified the template as 'app_form' not 'app_form.php'.  The docs will be updated soon.
